design/batch_retrieval_design.md

```diff
@@ -1,20 +1,176 @@
+
+ The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. See difference between proposed API and existing API see [Proposed API vs Old API](#proposed-api-vs-old-api) below.
+
+```
+POST /secrets/batch
+
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+```
+
+### Response
+
+All secret values will be treated as individual entities. If a single secret fails to be retrieved, this will not impact the integrity of the request.
+
+| Code             | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 200 OK           | All secret values were retrieved successfully                |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations) |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
+| 403 Forbidden    | The authenticated user lacks the necessary privilege         |
+| 404 Not Found    | At least one variable does not exist, or at least one variable does not have any secret values |
+
+If no Conjur/DAP secrets were able to be retrieved (no 200 code), the response will still return a 200 OK because the request body is parsable and request was handled.
+
+If an invalid request body was supplied, a 400 Bad Request will be returned
+
+#### Example response
+
+```
+HTTP/1.1 200 OK
+
+{
+  "successSecret1":[
+     {
+       "status": 200
+       "value": "superSecret"
+     }
+  ]
```

Comment:
I think the response body should be similar to the request, in case we want to add something to the response which is not specific to any of the variables:
```
{
  "variables": [
    {
      "name": "var/path/name1",
       "status": 200,
       "value": "superSecret"
    },
    {
      "name": "var/path/name2",
      "status": 404,
      "value": "Not Found"
    },
...
  ]
}
```

Reply:
Changed the body to resemble @orenbm comment [here](https://github.com/cyberark/secrets-provider-for-k8s/pull/205#discussion_r477111213)
```
{
  "successSecret1":{
    "status": 200,
    "value": "superSecret"
  },
  "failSecret1":{
    "status": 403,
    "value": "Forbidden"     
  }
```

Reply:
I can wrap it in variables object though

Reply:
I think your suggestion makes sense because it is possible that we will add filtering as part of the query param